tests/network/bindingplugin.go

```diff
@@ -0,0 +1,103 @@
+/*
+ * This file is part of the kubevirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/framework/kubevirt"
+	"kubevirt.io/kubevirt/tests/libkvconfig"
+	"kubevirt.io/kubevirt/tests/libvmi"
+	"kubevirt.io/kubevirt/tests/libwait"
+	"kubevirt.io/kubevirt/tests/testsuite"
+
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+)
+
+const (
+	passtBindingName  = "passt"
+	passtSidecarImage = "registry:5000/kubevirt/network-passt-binding:devel"
+	passtNetAttDef    = "netbindingpasst"
+	passtType         = "cni-passt-binding-plugin"
+)
+
+var _ = SIGDescribe("[Serial]network binding plugin", Serial, decorators.NetCustomBindingPlugins, func() {
+
+	libkvconfig.WithFeatureGate(virtconfig.NetworkBindingPlugingsGate)
```

Comment:
I find it odd to warp BeforeEach clause in a function, its not straight forward I had to dig in to realize it, I think its better to have them (BeforeEach clouses) inline.
Please consider inline it instead if its not mandatory.

Reply:
I dont recall they use it all over their tests, maybe for some specific tests. At least this is what I saw.

I would avoid hiding BeforeEach clauses and have the actual logic in a function, that way the structure remain and we have all boilerplate abstracted.
If there is more to the setup logic, tests would have many warpped BeforeEach calls instead of one.

Nothing prevent a BeforeEach from being called anywhere in a test container, at least we know what keyword to search to find it, having in wrapped makes it harder to find.

Reply:
k8s e2e test are using this on all their tests, it is the base of their framework [1].

I like the abstraction and boilerplate reduction, with a great opportunity to create focused resource management (e.g. change something and then revert it at teardown).

I expect a common test reader to never care how it is implemented. Just to know how to do something (e.g. enable a FG, change a parameter and revert it before completion, etc).
It makes the test smaller, more readable and much easier to improve the framework/tooling (because changes to the tooling has the potential not to change the tests).

[1] https://github.com/kubernetes/kubernetes/blob/6eee80fa9a41c015608a34dc6bc662a181f6ea43/test/e2e/framework/framework.go#L200